File: frontend/src/views/roadmap.ts
/**
 * Phase grouping for the "How to Improve" roadmap.
 *
 * Reads only fields already on a finding. No request, no new field, no LLM: the
 * function is pure, so the same findings always produce the same three buckets.
 *
 * ---------------------------------------------------------------------------- #
 * THE RULE
 *
 * The primary signal is `remediation_effort`, which the remediate stage already
 * writes for every open finding, and whose own prompt defines the three values as:
 *
 *   low    — "a configuration or design-document change"
 *   medium — "a component or flow change"
 *   high   — "a structural change to the architecture"
 *
 * Those definitions already are the three phases, so the rule leans on them rather
 * than inferring effort from severity, which measures how much a gap matters and
 * not how much work it is to close.
 *
 *   1. STRUCTURAL   if effort is "high", OR the fix touches more than one component.
 *   2. IMMEDIATE    else if effort is "low" AND severity is "high".
 *   3. SHORT-TERM   everything else.
 *
 * The component-count override in (1) exists because a change spanning three
 * components is a structural change whatever it was scored: the reviewer sequencing
 * work cares about blast radius, and that is the one effort signal in the data that
 * is not the model's opinion.
 *
 * Where `remediation_effort` is absent — an older stored review, or a check the
 * model omitted from its remediation list — the rule falls back to severity and
 * component count alone, which lands high-severity single-component gaps in
 * Immediate and the rest in Short-term. Absent effort never silently becomes "low".
 *
 * DELIBERATELY NOT USED: pillar. Routing Operational Excellence and Reliability to
 * Structural was considered and rejected — those pillars contain some of the
 * cheapest fixes in the rubric ("no runbook referenced", "health check interval not
 * specified") alongside genuinely structural ones ("single-AZ"). A pillar names a
 * topic, not an amount of work, so keying off it would file cheap documentation
 * fixes as architecture projects.
 * ---------------------------------------------------------------------------- #
 */

import type { Finding } from '../types'

export type Phase = 'immediate' | 'short_term' | 'structural'

/** Presentation order, and the order the phases are worked in. */
export const PHASE_ORDER: readonly Phase[] = ['immediate', 'short_term', 'structural']

export const PHASE_LABEL: Record<Phase, string> = {
  immediate: 'Immediate',
  short_term: 'Short-term',
  structural: 'Structural',
}

/** What each phase means, in the reviewer's terms rather than the rule's. */
export const PHASE_BLURB: Record<Phase, string> = {
  immediate: 'High-severity gaps closable by a configuration or document change.',
  short_term: 'A component or flow change — worth scheduling, not architecture work.',
  structural: 'Spans several components or needs a design change.',
}

/** Above one component, a fix is a structural change however it was scored. */
const STRUCTURAL_COMPONENT_COUNT = 2

export function phaseFor(finding: Finding): Phase {
  const spansComponents = finding.affected_components.length >= STRUCTURAL_COMPONENT_COUNT
  if (finding.remediation_effort === 'high' || spansComponents) return 'structural'
  if (finding.remediation_effort === 'low' && finding.severity === 'high') {
    return 'immediate'
  }
  if (finding.remediation_effort === '' && finding.severity === 'high') {
    // No effort recorded. Severity and blast radius are all there is, and a
    // single-component high-severity gap is the shape that is usually cheap.
    return 'immediate'
  }
  return 'short_term'
}

/**
 * Every open finding, in exactly one phase.
 *
 * A partition, not a shortlist: unlike Top Action Items there is no cap and no
 * dedupe, because this is the whole plan — a roadmap that silently dropped the
 * eleventh item would be worse than no roadmap.
 *
 * The same open/closed filter as everywhere else: a passing or not-applicable check
 * has nothing to schedule.
 */
export function groupByPhase(findings: readonly Finding[]): Record<Phase, Finding[]> {
  const grouped: Record<Phase, Finding[]> = {
    immediate: [],
    short_term: [],
    structural: [],
  }

  for (const finding of findings) {
    if (finding.status !== 'fail' && finding.status !== 'partial') continue
    grouped[phaseFor(finding)].push(finding)
  }

  for (const phase of PHASE_ORDER) grouped[phase].sort(byPriorityThenId)
  return grouped
}

/**
 * Remediation order, with an unranked 0 last rather than first — matching the
 * findings list and Top Action Items.
 *
 * The `check_id` tie-break is what makes the output independent of input order:
 * most findings in a real review share `priority: 0`, and a sort keyed on priority
 * alone would preserve whatever order the array happened to arrive in.
 */
function byPriorityThenId(a: Finding, b: Finding): number {
  const rank = (finding: Finding) =>
    finding.priority > 0 ? finding.priority : Number.MAX_SAFE_INTEGER
  if (rank(a) !== rank(b)) return rank(a) - rank(b)
  return `${a.framework}:${a.check_id}`.localeCompare(`${b.framework}:${b.check_id}`)
}

File: backend/roadmap.py
"""Phase grouping for the "How to Improve" roadmap.

The Python half of a rule that also lives in ``frontend/src/views/roadmap.ts``. The
two runtimes cannot share code, so ``fixtures/roadmap_cases.json`` is asserted by
both test suites: a change here that is not mirrored there fails, rather than
producing a PDF that quietly disagrees with the screen.

Pure and deterministic. Reads only fields already on a stored finding — no LLM call,
no new schema field, no stage re-run.

THE RULE
--------
The primary signal is ``remediation_effort``, already written by the remediate
stage, whose own prompt defines the values as:

    low    — "a configuration or design-document change"
    medium — "a component or flow change"
    high   — "a structural change to the architecture"

Those definitions already are the three phases, so the rule uses them rather than
inferring effort from severity, which measures how much a gap matters and not how
much work it is to close.

    1. STRUCTURAL    if effort is "high", OR the fix touches more than one component.
    2. UNSPECIFIED   else if effort is absent.
    3. IMMEDIATE     else if effort is "low" AND severity is "high".
    4. SHORT-TERM    everything else.

The component-count override exists because a change spanning three components is a
structural change whatever it was scored — blast radius is the one effort signal in
the data that is not the model's opinion. It is also why it outranks the absent case:
component count is MEASURED, so it still decides when the estimate is missing.

WHY ``unspecified`` EXISTS
--------------------------
Absent effort used to fall back to severity: high severity became IMMEDIATE, on the
reasoning that a single-component high-severity gap is usually cheap. That put
findings nobody had estimated into a bucket labelled "closable by a configuration or
document change" — a claim about how much work the fix is, made from data that does
not contain one. Everything else fell to SHORT-TERM, which claims "a component or
flow change" and is no better founded.

It mattered in practice. When the remediate stage returns nothing — which a real run
did, twice over — every finding has a blank effort, so the whole roadmap filed itself
as Immediate and Short-term work with confident phase blurbs and no estimate behind
any of it.

So a blank effort now lands in its own phase, ordered LAST because it is not a
priority claim in either direction, and labelled as the absence it is. This follows
the same rule the rest of the project does: an honest blank beats a fabricated value.
These findings are still ON the roadmap — dropping them would be worse, and
``group_by_phase`` is documented as a partition rather than a shortlist.

Pillar is deliberately NOT used: it names a topic, not an amount of work. See the
TypeScript twin for the worked example.
"""

from __future__ import annotations

from typing import Literal

from schema import Finding

Phase = Literal["immediate", "short_term", "structural", "unspecified"]

#: Presentation order. ``unspecified`` is last: it ranks nothing, it withholds a
#: ranking, so it must not sit among phases that assert one.
PHASE_ORDER: tuple[Phase, ...] = (
    "immediate",
    "short_term",
    "structural",
    "unspecified",
)

PHASE_LABEL: dict[Phase, str] = {
    "immediate": "Immediate",
    "short_term": "Short-term",
    "structural": "Structural",
    "unspecified": "Effort not estimated",
}

#: What each phase means, in the reviewer's terms rather than the rule's.
PHASE_BLURB: dict[Phase, str] = {
    "immediate": "High-severity gaps closable by a configuration or document change.",
    "short_term": "A component or flow change — worth scheduling, not architecture work.",
    "structural": "Spans several components or needs a design change.",
    "unspecified": (
        "No effort estimate came back for these, so they are not scheduled. That is "
        "not a judgement that they are cheap or expensive — it is the absence of one."
    ),
}

#: Above one component, a fix is a structural change however it was scored.
STRUCTURAL_COMPONENT_COUNT = 2

_OPEN_STATUSES = frozenset({"fail", "partial"})


def phase_for(finding: Finding) -> Phase:
    spans_components = len(finding.affected_components) >= STRUCTURAL_COMPONENT_COUNT
    if finding.remediation_effort == "high" or spans_components:
        return "structural"
    # Checked BEFORE the effort buckets, so an absent estimate cannot be inferred
    # into one. Blast radius is handled above and still wins, because it is measured.
    if finding.remediation_effort == "":
        return "unspecified"
    if finding.remediation_effort == "low" and finding.severity == "high":
        return "immediate"
    return "short_term"


def group_by_phase(findings: list[Finding]) -> dict[Phase, list[Finding]]:
    """Every open finding, in exactly one phase.

    A partition, not a shortlist: no cap and no dedupe, because this is the whole
    plan. A roadmap that silently dropped the eleventh item would be worse than no
    roadmap.
    """
    grouped: dict[Phase, list[Finding]] = {phase: [] for phase in PHASE_ORDER}
    for finding in findings:
        if finding.status not in _OPEN_STATUSES:
            continue
        grouped[phase_for(finding)].append(finding)

    for phase in PHASE_ORDER:
        grouped[phase].sort(key=_order)
    return grouped


def _order(finding: Finding) -> tuple[int, str]:
    """Remediation order, unranked last.

    The ``check_id`` tie-break is what makes the output independent of input order:
    most findings in a real review share ``priority == 0``, and a sort keyed on
    priority alone would preserve whatever order the list happened to arrive in.
    """
    rank = finding.priority if finding.priority > 0 else 2**31
    return (rank, f"{finding.framework}:{finding.check_id}")
